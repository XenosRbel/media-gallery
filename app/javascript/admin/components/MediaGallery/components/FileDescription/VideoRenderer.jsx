import React from "react";
import DocViewer from "react-doc-viewer";

const VideoRenderer = ({
  mainState: { currentDocument },
}) => {
  if (!currentDocument) return null;

  return (
    <div className="video-renderer" >
      <video controls>
        <source src={currentDocument.fileData}/>
      </video>
    </div>
  );
};

VideoRenderer.fileTypes = ["mp4", "video/mp4"];
VideoRenderer.weight = 1;

export default VideoRenderer;
