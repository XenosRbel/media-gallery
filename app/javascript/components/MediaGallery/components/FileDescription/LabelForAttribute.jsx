import React from 'react';
import PropTypes from 'prop-types';

import { sanitizeAttributeName } from '../../../common/utils';
import { translateFileAttribute } from '../../utils/common';

const LabelForAttribute = ({ name, value }) => (
  <div>
    <span variant="small">{translateFileAttribute(sanitizeAttributeName(name))}</span>
    <span>: </span>
    <span variant="small">{value}</span>
  </div>
);

LabelForAttribute.propTypes = {
  name: PropTypes.string,
  value: PropTypes.oneOfType([PropTypes.string, PropTypes.number]),
};

LabelForAttribute.defaultProps = {
  name: '',
  value: '',
};

export default LabelForAttribute;
