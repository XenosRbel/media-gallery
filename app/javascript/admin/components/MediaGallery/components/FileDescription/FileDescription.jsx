import React, { useState, useCallback, useEffect } from 'react';
import PropTypes from 'prop-types';
import { Row, Col, Button, message, Input, Upload } from 'antd';
import { CopyToClipboard } from 'react-copy-to-clipboard';

import InputField from '../../../common/inputs/InputField';
import { getInputProps, formatFileSize } from '../../../common/utils';
import { PUT_HTTP_METHOD, SUCCESS } from '../../constants/common';
import { fileShape } from '../../data/shapes';
import { translateFileAttribute, getErrorStatusByErrorCode, handleDeleteFilesByIds } from '../../utils/common';
import {
  TRANSLATE_FILE_DESCRIPTION_PLACEHOLDER,
  TRANSLATE_DELETE_ACTION,
  TRANSLATE_SAVE_ACTION,
  TRANSLATE_REPLACE_ACTION,
  ROW_GUTTER_SIZE,
  TRANSLATE_URL_COPIED,
  TRANSLATE_FILE_ATTACHED,
} from '../../constants/fileDescription';
import { buildFormData } from '../../utils/fileUploadForm';
import LabelForAttribute from './LabelForAttribute';
import { sendDataRequest } from '../../../../../front/utils/http';
import DocViewer, {
  BMPRenderer,
  HTMLRenderer,
  JPGRenderer,
  MSDocRenderer,
  MSGRenderer,
  PDFRenderer,
  PNGRenderer,
  TIFFRenderer,
  TXTRenderer
} from "react-doc-viewer";
import VideoRenderer from './VideoRenderer'

const showInfoMessage = message.info;
const onCopy = () => showInfoMessage(TRANSLATE_URL_COPIED);

function getBase64(file) {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();

    reader.readAsDataURL(file);
    reader.onload = () => resolve(reader.result);
    reader.onerror = (error) => reject(error);
  });
}

const FileDescription = ({
  galleryFile,
  filesPath,
  handleCloseFileDescription,
  handleLoadGalleryItems,
  handleResponseMessage,
}) => {
  const [file, setFile] = useState({});
  const [newFile, setNewFile] = useState(null);
  const [originalImage, setOriginalImage] = useState();

  const { id, name, folderName, width, height, filesize, url } = file;

  const handleAttributeChange = useCallback(({ target: { name: attribute, value } }) => {
    setFile((prevFile) => ({ ...prevFile, [attribute]: value }));
  }, []);

  const handleBeforeUpload = useCallback(async (currentNewFile) => {
    setNewFile(currentNewFile);

    setOriginalImage(await getBase64(currentNewFile));

    showInfoMessage(TRANSLATE_FILE_ATTACHED);

    return false;
  }, []);

  const handleSubmit = useCallback(async () => {
    const formData = buildFormData({ file: newFile, ...file });

    const data = await sendDataRequest(PUT_HTTP_METHOD, `${filesPath}/${id}`, formData);

    const requestStatus = getErrorStatusByErrorCode(data.status);

    if (requestStatus === SUCCESS) {
      handleCloseFileDescription();
      handleLoadGalleryItems();
    }

    handleResponseMessage(data);
  }, [file, filesPath, handleCloseFileDescription, handleResponseMessage, handleLoadGalleryItems, id, newFile]);

  const handleSubmitDeleteFile = useCallback(async () => {
    const { status } = await handleDeleteFilesByIds(filesPath, id, 'id');

    if (status === SUCCESS) {
      handleCloseFileDescription();
      handleLoadGalleryItems();
    }
  }, [filesPath, handleCloseFileDescription, handleLoadGalleryItems, id]);

  useEffect(() => {
    setFile(galleryFile);
    setOriginalImage(galleryFile.originalImage);
  }, [galleryFile]);

  return (
    <div className='file-description'>
      <DocViewer
        className="center"
        alt={name}
        pluginRenderers={[
          BMPRenderer,
          HTMLRenderer,
          JPGRenderer,
          MSDocRenderer,
          MSGRenderer,
          PDFRenderer,
          PNGRenderer,
          TIFFRenderer,
          TXTRenderer,
          VideoRenderer]}
        config={
          {
            header: {
             disableHeader: true,
             disableFileName: false,
             retainURLParams: false
            }
           }
        }
        documents={[ {uri: url }]}
      />

      <LabelForAttribute name="name" value={name} />

      <LabelForAttribute name="folderName" value={folderName} />

      <LabelForAttribute name="width" value={width} />

      <LabelForAttribute name="height" value={height} />

      <LabelForAttribute name="filesize" value={formatFileSize(filesize)} />

      <Input
        className="attribute"
        value={url}
        suffix={
          <CopyToClipboard onCopy={onCopy} text={url}>
            <span className="ant-btn ant-btn-icon-only">
              <i className="fa fa-clipboard copy-icon" aria-hidden="true" />
            </span>
          </CopyToClipboard>
        }
      />

      <form className="media-gallery-folder-form">
        <div>
          <InputField
            {...getInputProps('description', handleAttributeChange, file, translateFileAttribute)}
            className="ant-col attribute"
            placeholder={TRANSLATE_FILE_DESCRIPTION_PLACEHOLDER}
          />
        </div>
        <Row gutter={ROW_GUTTER_SIZE} className='actions'>
          <Col>
            <Button size="small" onClick={handleSubmitDeleteFile}>
              {TRANSLATE_DELETE_ACTION}
            </Button>
          </Col>
          <Col>
            <Upload showUploadList={false} beforeUpload={handleBeforeUpload} onChange={null}>
              <Button size="small">{TRANSLATE_REPLACE_ACTION}</Button>
            </Upload>
          </Col>
          <Col>
            <Button size="small" onClick={handleSubmit}>
              {TRANSLATE_SAVE_ACTION}
            </Button>
          </Col>
        </Row>
      </form>
    </div>
  );
};

FileDescription.propTypes = {
  filesPath: PropTypes.string.isRequired,
  galleryFile: PropTypes.shape(fileShape),
  handleCloseFileDescription: PropTypes.func.isRequired,
  handleLoadGalleryItems: PropTypes.func.isRequired,
  handleResponseMessage: PropTypes.func.isRequired,
};

FileDescription.defaultProps = {
  galleryFile: {
    description: '',
    filesize: 0,
    folderId: 0,
    folderName: '',
    height: 0,
    id: 0,
    name: '',
    originalImage: '',
    thumbnailImage: '',
    url: '',
    width: 0,
  }
};

export default FileDescription;
