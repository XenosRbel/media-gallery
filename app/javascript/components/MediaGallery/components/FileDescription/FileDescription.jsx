import React, { useState, useCallback, useEffect } from 'react';
import PropTypes from 'prop-types';
import { Row, Col, Button, message, Input, Upload } from 'antd';
import { CopyToClipboard } from 'react-copy-to-clipboard';

import InputField from '../../../inputs/InputField';
import { getInputProps, formatFileSize } from '../../../common/utils';
import { PUT_HTTP_METHOD, SUCCESS } from '../../constants/common';
import { fileShape } from '../../data/shapes';
import { translateFileAttribute, getErrorStatusByErrorCode, handleDeleteFilesByIds } from '../../utils/common';
import {
  TRANSLATE_FILE_DESCRIPTION_PLACEHOLDER,
  TRANSLATE_DELETE_ACTION,
  TRANSLATE_SAVE_ACTION,
  TRANSLATE_REPLACE_ACTION,
  ROW_GUTTER_SIZE,
  TRANSLATE_URL_COPIED,
  TRANSLATE_FILE_ATTACHED,
} from '../../constants/fileDescription';
import { buildFormData } from '../../utils/fileUploadForm';
import LabelForAttribute from './LabelForAttribute';
import { sendDataRequest } from '../../../common/http';

const showInfoMessage = message.info;
const onCopy = () => showInfoMessage(TRANSLATE_URL_COPIED);

function getBase64(file) {
  return new Promise((resolve, reject) => {
    const reader = new FileReader();

    reader.readAsDataURL(file);
    reader.onload = () => resolve(reader.result);
    reader.onerror = (error) => reject(error);
  });
}

const FileDescription = ({
  galleryFile,
  filesPath,
  handleCloseFileDescription,
  handleLoadGalleryItems,
  handleResponseMessage,
}) => {
  const [file, setFile] = useState({});
  const [newFile, setNewFile] = useState(null);
  const [originalImage, setOriginalImage] = useState();

  const { id, name, groupName, width, height, filesize, url } = file;

  const handleAttributeChange = useCallback(({ target: { name: attribute, value } }) => {
    setFile((prevFile) => ({ ...prevFile, [attribute]: value }));
  }, []);

  const handleBeforeUpload = useCallback(async (currentNewFile) => {
    setNewFile(currentNewFile);

    setOriginalImage(await getBase64(currentNewFile));

    showInfoMessage(TRANSLATE_FILE_ATTACHED);

    return false;
  }, []);

  const handleSubmit = useCallback(async () => {
    const formData = buildFormData({ file: newFile, ...file });

    const data = await sendDataRequest(PUT_HTTP_METHOD, `${filesPath}/${id}`, formData);

    const requestStatus = getErrorStatusByErrorCode(data.status);

    if (requestStatus === SUCCESS) {
      handleCloseFileDescription();
      handleLoadGalleryItems();
    }

    handleResponseMessage(data);
  }, [file, filesPath, handleCloseFileDescription, handleResponseMessage, handleLoadGalleryItems, id, newFile]);

  const handleSubmitDeleteFile = useCallback(async () => {
    const { status } = await handleDeleteFilesByIds(filesPath, id, 'id');

    if (status === SUCCESS) {
      handleCloseFileDescription();
      handleLoadGalleryItems();
    }
  }, [filesPath, handleCloseFileDescription, handleLoadGalleryItems, id]);

  useEffect(() => {
    setFile(galleryFile);
    setOriginalImage(galleryFile.originalImage);
  }, [galleryFile]);

  return (
    <div>
      <img className="center" src={originalImage} alt={name} />

      <LabelForAttribute name="name" value={name} />

      <LabelForAttribute name="groupName" value={groupName} />

      <LabelForAttribute name="width" value={width} />

      <LabelForAttribute name="height" value={height} />

      <LabelForAttribute name="filesize" value={formatFileSize(filesize)} />

      <Input
        value={url}
        suffix={
          <CopyToClipboard onCopy={onCopy} text={url}>
            <span className="antd-btn antd-btn-icon-only">
              <i className="fa fa-clipboard copy-icon" aria-hidden="true" />
            </span>
          </CopyToClipboard>
        }
      />

      <form className="media-gallery-group-form">
        <div>
          <InputField
            {...getInputProps('description', handleAttributeChange, file, translateFileAttribute)}
            className="antd-col"
            placeholder={TRANSLATE_FILE_DESCRIPTION_PLACEHOLDER}
          />
        </div>
        <Row gutter={ROW_GUTTER_SIZE}>
          <Col>
            <Button size="small" onClick={handleSubmitDeleteFile}>
              {TRANSLATE_DELETE_ACTION}
            </Button>
          </Col>
          <Col>
            <Upload showUploadList={false} beforeUpload={handleBeforeUpload} onChange={null}>
              <Button size="small">{TRANSLATE_REPLACE_ACTION}</Button>
            </Upload>
          </Col>
          <Col>
            <Button size="small" onClick={handleSubmit}>
              {TRANSLATE_SAVE_ACTION}
            </Button>
          </Col>
        </Row>
      </form>
    </div>
  );
};

FileDescription.propTypes = {
  filesPath: PropTypes.string.isRequired,
  galleryFile: PropTypes.shape(fileShape).isRequired,
  handleCloseFileDescription: PropTypes.func.isRequired,
  handleLoadGalleryItems: PropTypes.func.isRequired,
  handleResponseMessage: PropTypes.func.isRequired,
};

export default FileDescription;
