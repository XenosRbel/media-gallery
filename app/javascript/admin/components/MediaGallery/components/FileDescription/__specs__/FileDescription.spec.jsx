import React from 'react';
import { shallow } from 'enzyme';

import FileDescription from '../FileDescription';

describe('Component: FileDescription', () => {
  const file = {
    description: '',
    filesize: 1024,
    folderId: 1,
    height: 1,
    id: 1,
    name: '',
    thumbnailImage: '',
    url: '',
    width: 1,
  };

  const filesPath = 'http://test.com';
  const handleCloseFileDescription = jest.fn();
  const handleLoadGalleryItems = jest.fn();
  const handleResponseMessage = jest.fn();

  const wrapper = shallow(
    <FileDescription
      galleryFile={file}
      filesPath={filesPath}
      handleCloseFileDescription={handleCloseFileDescription}
      handleLoadGalleryItems={handleLoadGalleryItems}
      handleResponseMessage={handleResponseMessage}
    />,
  );

  it('renders correctly', () => {
    expect(wrapper).toMatchSnapshot();
  });
});
