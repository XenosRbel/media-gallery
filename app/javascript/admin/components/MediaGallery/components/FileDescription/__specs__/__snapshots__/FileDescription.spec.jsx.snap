// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Component: FileDescription renders correctly 1`] = `
<div>
  <img
    className="center"
  />
  <LabelForAttribute
    name="name"
    value=""
  />
  <LabelForAttribute
    name="folderName"
    value=""
  />
  <LabelForAttribute
    name="width"
    value=""
  />
  <LabelForAttribute
    name="height"
    value=""
  />
  <LabelForAttribute
    name="filesize"
    value="NaN undefined"
  />
  <ForwardRef
    onBlur={[Function]}
    onChange={[Function]}
    onFocus={[Function]}
    suffix={
      <CopyToClipboard
        onCopy={[Function]}
      >
        <span
          className="ant-btn ant-btn-icon-only"
        >
          <i
            aria-hidden="true"
            className="fa fa-clipboard copy-icon"
          />
        </span>
      </CopyToClipboard>
    }
    theme="default"
  />
  <form
    className="media-gallery-folder-form"
  >
    <div>
      <InputField
        className="ant-col"
        disabled={false}
        id="descriptionInput"
        label={true}
        name="description"
        onChange={[Function]}
        placeholder="i18n_mock"
        translateAttribute={[Function]}
        value=""
      />
    </div>
    <Row
      gutter={
        Array [
          24,
          8,
        ]
      }
    >
      <Col>
        <n
          block={false}
          disabled={false}
          ghost={false}
          htmlType="button"
          icon={null}
          loading={false}
          onClick={[Function]}
          shape="round"
          size="small"
          styles=""
          target=""
          theme="default"
          type="primary"
          wrapperClassName=""
        >
          i18n_mock
        </n>
      </Col>
      <Col>
        <Upload
          accept=""
          action=""
          beforeUpload={[Function]}
          className=""
          data={Object {}}
          disabled={false}
          listType="text"
          multiple={false}
          onChange={null}
          showUploadList={false}
          supportServerRender={true}
          type="select"
        >
          <n
            block={false}
            disabled={false}
            ghost={false}
            htmlType="button"
            icon={null}
            loading={false}
            onClick={[Function]}
            shape="round"
            size="small"
            styles=""
            target=""
            theme="default"
            type="primary"
            wrapperClassName=""
          >
            i18n_mock
          </n>
        </Upload>
      </Col>
      <Col>
        <n
          block={false}
          disabled={false}
          ghost={false}
          htmlType="button"
          icon={null}
          loading={false}
          onClick={[Function]}
          shape="round"
          size="small"
          styles=""
          target=""
          theme="default"
          type="primary"
          wrapperClassName=""
        >
          i18n_mock
        </n>
      </Col>
    </Row>
  </form>
</div>
`;
